src/mlpack/methods/ann/activation_functions/silu_function.hpp

```diff
@@ -68,26 +68,36 @@ class SILUFunction
   /**
    * Computes the first derivative of the SILU function.
    *
-   * @param y Input activation.
+   * @param x Input activation.
+   * @param y Result of Fn(x).
    * @return f'(x)
    */
-  static double Deriv(const double x)
+  static double Deriv(const double x, const double y)
   {
-    double sigmoid = 1.0 / (1.0 + std::exp(-x));
-    return sigmoid * (1.0 + x * (1.0 - sigmoid));
+    // since y = x * sigmoid(x)
+    double sigmoid = y / x; // save an exp
+    return x == 0 ? 0.5 : sigmoid * (1.0 + x * (1.0 - sigmoid));
+    // the expression above is indeterminate at 0, even though
+    // the expression solely in terms of x is defined (= 0.5)
   }
 
   /**
    * Computes the first derivatives of the SILU function.
    *
-   * @param y Input activations.
-   * @param x The resulting derivatives.
+   * @param x Input activation.
+   * @param y Result of Fn(x).
+   * @param dy The resulting derivatives.
    */
-  template<typename InputVecType, typename OutputVecType>
-  static void Deriv(const InputVecType &x, OutputVecType &y)
+  template<typename InputVecType, typename OutputVecType, typename DerivVecType>
+  static void Deriv(const InputVecType &x, const OutputVecType &y, DerivVecType &dy)
   {
-    OutputVecType sigmoid = 1.0 / (1.0 + arma::exp(-x));
-    y = sigmoid % (1.0 + x % (1.0 - sigmoid));
+    // since y = x * sigmoid(x)
+    // DerivVecType sigmoid = y / x;
+    // dy = sigmoid % (1.0 + x % (1.0 - sigmoid));
+    dy = (y / x) % (1.0 + x - y);
```

Comment:
```suggestion
    dy = (x != 0) % (y / x) % (1.0 + x - y) + (x == 0) % 0.5;
```
Maybe this helps clean up the lines below?  However, I'm not convinced which way is faster with Armadillo.  It would probably need some timing.  Up to you.

Reply:
actually, looks like this doesn't work as is.  would have to do `(x == 0) % arma::ones<MatType>(size(x)) % 0.5`
might still be ok/as good, but not positive.  still cleaner? or should I put it back the way i did it originally?

Reply:
I think it's fine as-is... no need to guess at what's faster, we can come back some other time and optimize it as needed.  The `dy(arma::find(...))` strikes me as something that is going to be slow, but glancing at Armadillo sources it seems like that might actually be quite optimized, and I think it avoids materializing the result of the `find(x == 0)` expression, which is quite clever!

Reply:
Yeh, I tested it out and it seemed to be pretty fast.  I have used that a lot in other code to replace nans, and haven't found a better way.